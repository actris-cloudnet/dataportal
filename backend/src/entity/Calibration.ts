import { Column, Entity, BeforeInsert, BeforeUpdate } from "typeorm";

@Entity()
export class Calibration {
  @Column({ primary: true })
  instrumentPid!: string;

  @Column({ type: "date", primary: true })
  measurementDate!: string;

  @Column({ type: "jsonb" })
  data!: any;

  @Column()
  createdAt!: Date;

  @Column()
  updatedAt!: Date;

  @BeforeInsert()
  updateDateCreation() {
    this.createdAt = new Date();
    this.updatedAt = this.createdAt;
  }

  @BeforeUpdate()
  updateDateUpdate() {
    this.updatedAt = new Date();
  }
}
