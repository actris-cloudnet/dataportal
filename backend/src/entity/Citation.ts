import {Column, Entity, ManyToMany, ManyToOne, PrimaryColumn, PrimaryGeneratedColumn} from 'typeorm'
import {Site} from './Site'
import {Model} from './Model'


@Entity()
export class Citation {

  @PrimaryColumn()
  id!: string

  @Column({type: 'text'})
  acknowledgements!: string

}

@Entity()
export class RegularCitation extends Citation {

  @ManyToMany(_ => Site)
  sites!: Site[]
}

@Entity()
export class ModelCitation extends Citation {

  @ManyToMany(_ => Model)
  models!: Model[]
}
