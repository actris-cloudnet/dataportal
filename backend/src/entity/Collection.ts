import {
  BeforeInsert,
  BeforeUpdate,
  Column,
  Entity,
  JoinTable,
  ManyToMany,
  PrimaryColumn,
} from 'typeorm'
import {File} from './File'
import { v4 as generateUuidV4 } from 'uuid'

@Entity()
export class Collection {

  @PrimaryColumn('uuid')
  uuid!: string

  @ManyToMany(_type => File)
  @JoinTable()
  files!: File[]

  @Column({default: ''})
  title!: string

  @Column({default: ''})
  pid!: string

  @Column({default: 0})
  downloadCount!: number

  @Column()
  createdAt!: Date

  @Column()
  updatedAt!: Date

  @BeforeInsert()
  setCreatedAt() {
    this.createdAt = new Date()
    this.updatedAt = this.createdAt
  }

  @BeforeUpdate()
  setUpdatedAt() {
    this.updatedAt = new Date()
  }

  constructor(files: File[]) {
    this.uuid = generateUuidV4()
    this.files = files
  }
}
