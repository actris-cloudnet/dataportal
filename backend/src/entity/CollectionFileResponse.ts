import {File, ModelFile, RegularFile} from './File'

export class CollectionFileResponse {

  uuid: string
  measurementDate: Date
  site: string
  siteId: string
  product: string
  productId: string
  modelId: string | null
  size: number
  volatile: boolean
  legacy: boolean

  constructor(file: RegularFile | ModelFile) {
    this.uuid = file.uuid
    this.measurementDate = file.measurementDate
    this.site = file.site.humanReadableName
    this.siteId = file.site.id
    this.product = file.product.humanReadableName
    this.productId = file.product.id
    this.modelId = 'model' in file ? file.model.id : null
    this.size = file.size
    this.volatile = file.volatile
    this.legacy = file.legacy
  }
}
