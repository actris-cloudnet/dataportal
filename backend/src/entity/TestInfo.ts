import { Column, Entity, PrimaryColumn } from "typeorm";

@Entity()
export class TestInfo {
  @PrimaryColumn({ type: "text" })
  testId!: string;

  @Column({ type: "text" })
  name!: string;

  @Column({ type: "text" })
  description!: string;
}
