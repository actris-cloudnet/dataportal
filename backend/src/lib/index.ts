export const stringify = (obj: any): string => JSON.stringify(obj, null, 2)
