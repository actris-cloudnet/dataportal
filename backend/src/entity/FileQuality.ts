import { Column, Entity, OneToMany, PrimaryColumn } from "typeorm";
import { QualityReport } from "./QualityReport";

export enum ErrorLevel {
  PASS = "pass",
  WARNING = "warning",
  ERROR = "error",
}

@Entity()
export class FileQuality {
  @PrimaryColumn("uuid")
  uuid!: string;

  @Column({
    type: "enum",
    enum: ErrorLevel,
  })
  errorLevel!: ErrorLevel;

  @Column({ type: "text" })
  qcVersion!: string;

  @Column()
  timestamp!: Date;

  @Column({ type: "smallint", nullable: true })
  tests!: number | null;

  @Column({ type: "smallint", nullable: true })
  errors!: number | null;

  @Column({ type: "smallint", nullable: true })
  warnings!: number | null;

  @OneToMany(() => QualityReport, (report) => report.quality)
  testReports!: QualityReport[];
}
