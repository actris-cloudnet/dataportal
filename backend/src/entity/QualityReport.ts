import { Column, Entity, ManyToOne, PrimaryColumn } from "typeorm";
import { FileQuality } from "./FileQuality";

export enum ErrorLevel {
  PASS = "pass",
  WARNING = "warning",
  ERROR = "error",
}

@Entity()
export class QualityReport {
  @PrimaryColumn({ type: "text" })
  testId!: string;

  @Column({ type: "text" })
  description!: string;

  @Column({
    type: "enum",
    enum: ErrorLevel,
  })
  result!: ErrorLevel;

  @Column({ type: "jsonb", nullable: true })
  exceptions!: any;

  @ManyToOne(() => FileQuality, (quality) => quality.testReports, { primary: true, onDelete: "CASCADE" })
  quality!: FileQuality;
}
