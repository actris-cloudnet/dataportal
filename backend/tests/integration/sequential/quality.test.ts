import { Connection, createConnection, Repository } from "typeorm";
import { QualityReport } from "../../../src/entity/QualityReport";
import axios from "axios";
import { backendPrivateUrl, backendPublicUrl } from "../../lib";
import { promises as fsp } from "fs";
import { FileQuality } from "../../../src/entity/FileQuality";
import { readResources } from "../../../../shared/lib";
import { TestInfo } from "../../../src/entity/TestInfo";

let conn: Connection;
let testInfoRepo: Repository<TestInfo>;
let qualityReportRepo: Repository<QualityReport>;
let fileQualityRepo: Repository<FileQuality>;
const privateUrl = `${backendPrivateUrl}quality/`;
const publicUrl = `${backendPublicUrl}quality/`;
const fileUrl = `${backendPublicUrl}files/`;
const searchUrl = `${backendPublicUrl}search/`;
let resources: any;
let report: any;

beforeAll(async () => {
  resources = await readResources();
  conn = await createConnection();
  testInfoRepo = await conn.getRepository("test_info");
  qualityReportRepo = await conn.getRepository("quality_report");
  fileQualityRepo = await conn.getRepository("file_quality");
});

beforeEach(async () => {
  await cleanRepos();
  await conn
    .getRepository("regular_file")
    .save(JSON.parse((await fsp.readFile("fixtures/2-regular_file.json")).toString()));
  await conn
    .getRepository("model_file")
    .save(JSON.parse((await fsp.readFile("fixtures/2-model_file.json")).toString()));
  await conn
    .getRepository("search_file")
    .save(JSON.parse((await fsp.readFile("fixtures/2-search_file.json")).toString()));
  await testInfoRepo.save(JSON.parse((await fsp.readFile("fixtures/0-test_info.json")).toString()));
});

afterAll(async () => {
  await cleanRepos();
  await conn.close();
});

describe("PUT /quality/:uuid", () => {
  it("fails with 400 if file does not exist", async () => {
    await expect(axios.put(`${privateUrl}4FC4577C-84BF-4557-86D8-1A1FB8D1D81E`, report)).rejects.toMatchObject({
      response: { status: 400 },
    });
  });

  it("creates new report for model file (no prior report)", async () => {
    const res1 = await axios.get(searchUrl, { params: { dateFrom: "2020-12-05" } });
    expect(res1.data[1].errorLevel).toEqual(null);
    const uuid = "b5d1d5af-3667-41bc-b952-e684f627d91c";
    await putReportWithErrors(uuid);
    const res = await axios.get(searchUrl, { params: { dateFrom: "2020-12-05" } });
    expect(res.data[1].errorLevel).toEqual("error");
    // test that search_file table is updated
    const res2 = await axios.get(searchUrl, { params: { dateFrom: "2020-12-05" } });
    expect(res2.data[0].errorLevel).toEqual("error");
  });

  it("creates report for regular file (old report exist)", async () => {
    const uuid = "acf78456-11b1-41a6-b2de-aa7590a75675";
    await putReportWithErrors(uuid);
    const response = await axios.get(`${publicUrl}${uuid}`);
    expect(response.status).toEqual(200);
    expect(response.data).toMatchSnapshot();
  });

  it("works with warnings only report", async () => {
    const uuid = "acf78456-11b1-41a6-b2de-aa7590a75675";
    await putReportWithWarnings(uuid);
    const response = await axios.get(`${publicUrl}${uuid}`);
    expect(response.status).toEqual(200);
    expect(response.data).toMatchSnapshot();
  });

  it("works with passing report", async () => {
    const uuid = "acf78456-11b1-41a6-b2de-aa7590a75675";
    await putReportWithPasses(uuid);
    const response = await axios.get(`${publicUrl}${uuid}`);
    expect(response.status).toEqual(200);
    expect(response.data).toMatchSnapshot();
  });

  it("doesn't update file's updatedAt when creating or updating report", async () => {
    await expect(axios.get(`${fileUrl}acf78456-11b1-41a6-b2de-aa7590a75675`)).resolves.toMatchObject({
      status: 200,
      data: { updatedAt: "2021-02-22T10:39:58.449Z" },
    });
    await expect(axios.put(`${privateUrl}acf78456-11b1-41a6-b2de-aa7590a75675`, report)).resolves.toMatchObject({
      status: 201,
    });
    await expect(axios.get(`${fileUrl}acf78456-11b1-41a6-b2de-aa7590a75675`)).resolves.toMatchObject({
      status: 200,
      data: { updatedAt: "2021-02-22T10:39:58.449Z" },
    });
  });
});

async function putReportWithErrors(uuid: string) {
  report = resources["quality-report-error"];
  await expect(axios.put(`${privateUrl}${uuid}`, report)).resolves.toMatchObject({
    status: 201,
  });
}

async function putReportWithWarnings(uuid: string) {
  report = resources["quality-report-warning"];
  await expect(axios.put(`${privateUrl}${uuid}`, report)).resolves.toMatchObject({
    status: 201,
  });
}

async function putReportWithPasses(uuid: string) {
  report = resources["quality-report-pass"];
  await expect(axios.put(`${privateUrl}${uuid}`, report)).resolves.toMatchObject({
    status: 201,
  });
}

async function cleanRepos() {
  await qualityReportRepo.delete({});
  await fileQualityRepo.delete({});
  await conn.getRepository("visualization").delete({});
  await conn.getRepository("model_visualization").delete({});
  await conn.getRepository("regular_file").delete({});
  await conn.getRepository("model_file").delete({});
  await conn.getRepository("search_file").delete({});
  return;
}
