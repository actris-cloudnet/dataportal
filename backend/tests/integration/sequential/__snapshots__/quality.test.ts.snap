// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PUT /quality/:uuid creates report for regular file (old report exist) 1`] = `
Object {
  "errorLevel": "error",
  "errors": 1,
  "qcVersion": "1.1.2",
  "testReports": Array [
    Object {
      "description": null,
      "exceptions": Array [
        Object {
          "message": "Instrument PID is missing.",
          "result": "error",
        },
      ],
      "name": "TestInstrumentPid",
      "result": "warning",
      "testId": "TestInstrumentPid",
    },
    Object {
      "description": "Description for TestUnits",
      "exceptions": Array [],
      "name": "Name for TestUnits",
      "result": "pass",
      "testId": "TestUnits",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestCFConvention",
      "result": "pass",
      "testId": "TestCFConvention",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestTimeVector",
      "result": "pass",
      "testId": "TestTimeVector",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestVariableNames",
      "result": "pass",
      "testId": "TestVariableNames",
    },
  ],
  "tests": 6,
  "timestamp": "2022-10-13T07:00:26.906Z",
  "uuid": "acf78456-11b1-41a6-b2de-aa7590a75675",
  "warnings": 1,
}
`;

exports[`PUT /quality/:uuid works with passing report 1`] = `
Object {
  "errorLevel": "pass",
  "errors": 0,
  "qcVersion": "1.1.2",
  "testReports": Array [
    Object {
      "description": "Description for TestUnits",
      "exceptions": Array [],
      "name": "Name for TestUnits",
      "result": "pass",
      "testId": "TestUnits",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestCFConvention",
      "result": "pass",
      "testId": "TestCFConvention",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestInstrumentPid",
      "result": "pass",
      "testId": "TestInstrumentPid",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestTimeVector",
      "result": "pass",
      "testId": "TestTimeVector",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestVariableNames",
      "result": "pass",
      "testId": "TestVariableNames",
    },
  ],
  "tests": 5,
  "timestamp": "2022-10-13T07:00:26.906Z",
  "uuid": "acf78456-11b1-41a6-b2de-aa7590a75675",
  "warnings": 0,
}
`;

exports[`PUT /quality/:uuid works with warnings only report 1`] = `
Object {
  "errorLevel": "warning",
  "errors": 0,
  "qcVersion": "1.1.2",
  "testReports": Array [
    Object {
      "description": null,
      "exceptions": Array [
        Object {
          "message": "Instrument PID is missing.",
          "result": "warning",
        },
      ],
      "name": "TestInstrumentPid",
      "result": "warning",
      "testId": "TestInstrumentPid",
    },
    Object {
      "description": "Description for TestUnits",
      "exceptions": Array [],
      "name": "Name for TestUnits",
      "result": "pass",
      "testId": "TestUnits",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestCFConvention",
      "result": "pass",
      "testId": "TestCFConvention",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestTimeVector",
      "result": "pass",
      "testId": "TestTimeVector",
    },
    Object {
      "description": null,
      "exceptions": Array [],
      "name": "TestVariableNames",
      "result": "pass",
      "testId": "TestVariableNames",
    },
  ],
  "tests": 5,
  "timestamp": "2022-10-13T07:00:26.906Z",
  "uuid": "acf78456-11b1-41a6-b2de-aa7590a75675",
  "warnings": 1,
}
`;
